### DISCUSSION TITLE
Eventcatcher: Why do we need to listen on the created domNode?

### DISCUSSION THREAD

**Discussion by BurningTreeC** [TiddlyWiki/TiddlyWiki5#6669] [Ideas]
I'd like to discuss here if it would make sense to alternatively listen on a specified domNode (through a selector) for the events

Like that, we can place our eventcatcher somewhere in the widget tree and listen for events on a parent domNode for example

That comes in handy if we need access to variables defined somewhere in the widget tree (like let's say in the current tiddler) AND we maybe want the mousemove coordinates from the .tc-page-container (so that we have global mousemove event-variables)

I give that as an example because I'm trying to create a tiddler-drag-and drop solution using the eventcatcher and listening on the tiddler for the mousemove event causes it to interrupt if we move the mouse faster so that it goes out of the boundaries of the tiddler. I need to place the eventcatcher widget inside the tiddler's view- and edit-templates because I need access to variables defined there. It should be possible to drag more tiddlers at once so I define a qualified state tiddler based on the tiddler-title and a qualified state.

The change needed would be very simple:

- an attribute that accepts a selector for the domNode where the listener(s) get attached to
- a `querySelector` for that domNode
- if that domNode exists attach the eventlistener to it, otherwise to the eventcatcher-domNode

What do you think?

**Comment by saqimtiaz** [comment]
The problem you are going to run into is the same as with using pointer capture, a refresh cycle can destroy and recreate DOM nodes. So any attempt at attaching event listeners to DOM nodes not created by the widget assigning the listeners is going to be brittle.

> I need to place the eventcatcher widget inside the tiddler's view- and edit-templates because I need access to variables defined there.

Make those variables available to eventcatcher actions by setting up appropriate DOM node attributes.

**Reply by BurningTreeC** [reply]
Thank you @saqimtiaz ,

I now understand the problem.

> Make those variables available to eventcatcher actions by setting up appropriate DOM node attributes.

I don't know how to do that yet but I'll maybe get an idea

Do you think it's possible somehow listening for the touchmove event on the `.tc-page-container` and somehow getting the coordinates for multiple touches?

**Reply by saqimtiaz** [reply]
> I don't know how to do that yet but I'll maybe get an idea

Here is an example:

https://github.com/saqimtiaz/streams/blob/17dfc9cf634338f986acc99413113884cfaba2a0/plugins/streams/templates/stream-row-template.tid#L18-L19

In the eventcatcher actions, those DOM node attributes are available as variables prefixed with `dom-`.

**Reply by BurningTreeC** [reply]
Thank you @saqimtiaz ,

what do you think about listening on a different DOM node WITHIN the eventcatcher widget?

Like:

```
var selectorDomNode = domNode.querySelector(this.getAttribute("domNodeSelector")) || domNode;

...

selectorDomNode.addEventListener ...

```

**Reply by saqimtiaz** [reply]
Same problem with the refresh cycle applies.
